.gitignore

```diff
@@ -1,5 +1,5 @@
 .DS_Store
-__pycache__
+**__pycache__

```

Comment:
gitignore rules are weird, but I think the standard way to ignore a file or directory from any directory route is:
```
**/*/__pycache__
```

Reply:
Re your comment:

> interesting, and definitely something I should keep in mind
> 
> for this particular case, how would you overcome the fact the set of
> possible incorrect answers are currently generated  stochastically when the
> question is generated? i.e., even when I dictate the correct game and the
> question template, I won't know what the alternative answers are, by
> design? would you mock answer generation in test so that the alternative
> answers are static? would you ignore the constraint that in practice, all
> answers will be from a given set of options?

Yeah – this is part of the reason I'd avoid implementing the feedback, because it'd require you to go all the way back up to your most basic test fixtures, which is maybe a plan for a day when you want to move laterally instead of forward.

Basically, I think instead of using sample data for your fixtures, you should use generated data. Those generated fixtures should be exemplary of the variety existent in your production data. So you'd have something like `test_question  is a function of (test_correct_answer, test_incorrect_answer)`.

If you want to verify that invoked production code random in some way, then I would supply a mock random function to the unit under test.

Does that make sense?

One thing I also forgot to mention about using random test fixtures: sometimes I do use random as evidence that I mean "any input of a given type has the following behavior" instead of "this specific input has the following behavior".

There is a whole class of tests that generate random inputs to your tests (popular Python package: https://pypi.org/project/hypothesis/) in hopes that it will find one that breaks your code. I've never used it in a production capacity – it seems like a pretty big sledgehammer. The nice thing about hypothesis is that it carries state around from test run to test run such that if/when a generated input fails, that same input will always be provided on future test runs to ensure there is no regression around the class of bug it caught the first time.